lib/internal/modules/cjs/loader.js

```diff
@@ -216,6 +216,13 @@ function tryExtensions(p, exts, isMain) {
   return false;
 }
 
+function readExtensions() {
+  const exts = Object.keys(Module._extensions);
+  for (var i = 0; i < exts.length; ++i)
+    exts[i] = path.extname(exts[i]) || exts[i];
+  return exts;
+}
```

Comment:
In the example case of `require.extensions` of `.foo.bar` and `.bar` this will return an array of `['.bar', '.bar]`?
So if you `delete require.extensions[".bar"]` files like `test-extensions.foo.bar` will still be found but their compiler at `require.extensions[".foo.bar"]` won't be used for the module since `_compile` will use `extname(filename)` to and get `.bar` which doesn't exist on `require.extensions` and so will default to `.js`.